h/streamer/messages.py

```diff
@@ -79,15 +76,43 @@ def handle_message(message, session, topic_handlers):
     # to stop connections being added or dropped during iteration, and if that
     # happens Python will throw a "Set changed size during iteration" error.
     sockets = list(websocket.WebSocket.instances)
+    handler(message.payload, sockets, session)
+
+
+def handle_annotation_event(message, sockets, session):
+    id_ = message['annotation_id']
+    annotation = storage.fetch_annotation(session, id_)
+
+    # FIXME: It isn't really nice to try and get the userid from the fetched
+    # annotation or otherwise get it from the maybe-already serialized
+    # annotation dict, to then only access the database for the nipsa flag once.
+    # We do this because the event action is `delete` at which point we can't
+    # load the annotation from the database. Handling annotation deletions is
+    # a known problem and will be fixed in the future.
+    userid = None
+    if annotation:
+        userid = annotation.userid
+    else:
+        userid = message.get('annotation_dict', {}).get('userid')
+    nipsa_service = NipsaService(session)
+    user_nipsad = nipsa_service.is_flagged(userid)
+
     for socket in sockets:
-        reply = handler(data, socket, session)
+        reply = _generate_annotation_event(message, socket, annotation, user_nipsad)
         if reply is None:
             continue
-        if not socket.terminated:
-            socket.send(json.dumps(reply))
+        socket.send_json(reply)
 
 
-def handle_annotation_event(message, socket, session):
+def handle_user_event(message, sockets, _):
+    for socket in sockets:
+        reply = _generate_user_event(message, socket)
+        if reply is None:
+            continue
+        socket.send_json(reply)
+
+
+def _generate_annotation_event(message, socket, annotation, user_nipsad):
```

Comment:
By way of a general observation rather than feedback on this PR specifically:

I think we could clean this up some more by realising that some of the conditions in this function refer only to the message, while some are about the (message, socket) pair. All of the conditions which are solely about the message could move to the top level, outside the socket loop.